qiskit/primitives/fidelity/fidelity.py

```diff
@@ -0,0 +1,97 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Zero probability fidelity primitive
+"""
+from typing import Callable, List, Union, Optional
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.primitives import Sampler
+from .base_fidelity import BaseFidelity
+
+
+SamplerFactory = Callable[[List[QuantumCircuit]], Sampler]
+
+
+class Fidelity(BaseFidelity):
+    """
+    Calculates the fidelity of two quantum circuits by measuring the zero probability outcome.
+    """
+
+    def __init__(
+        self,
+        left_circuit: QuantumCircuit,
+        right_circuit: QuantumCircuit,
+        sampler_factory: SamplerFactory,
+    ) -> None:
+        r"""
+        Initializes the class to evaluate the fidelities defined as the state overlap
+            :math:`|\langle\psi(x)|\phi(y)\rangle|^2`,
+        where x and y are parametrizations of the circuits :math:`\psi` and :math:`\phi`.
+        Args:
+            left_circuit: (Parametrized) quantum circuit :math:`|\psi\rangle`
+            right_circuit: (Parametrized) quantum circuit :math:`|\phi\rangle`
+            sampler_factory: Partial sampler used as a backend
+        Raises:
+            ValueError: left_circuit and right_circuit don't have the same number of qubits
+        """
+        super().__init__(left_circuit, right_circuit)
+
+        circuit = self._left_circuit.compose(self._right_circuit.inverse())
+        circuit.measure_all()
+
+        self.sampler = sampler_factory([circuit])
```

Comment:
That's a question and not a comment: is there a best practice as to which statements should go into `__init__` and which into `__enter__`? Intuitively it would seem that creating the sampler sessions is something that would go into `__enter__`, but I know that we also don't do that in the sampler itself... 🤔